Web language selection falls back for underscore locale codes after dynamic-language update

### I have searched the existing issues, both open and closed, to make sure this is not a duplicate report.

- [x] Yes

### The bug

After `feat: dynamic languages` (#27869), the web language-selection flow no longer round-trips several valid underscore-based locale codes that already exist in `i18n/*.json`.

Representative affected codes:
- `de_CH`
- - `pt_BR`
- - - `sr_Cyrl`
- - - - `zh_Hant`
The current web code keeps underscore codes in `langs`, but compares the selected/stored language against a hyphen-normalized `langCodes` list. Because of that mismatch, `getClosestAvailableLocale([$lang], langCodes)` returns `undefined` for those stored values, and the selector falls back to English instead of resolving the supported locale.

### The OS that Immich Server is running on

Ubuntu 24.04 (local development checkout)

### Version of Immich Server

main (validated on current upstream/main on 2026-04-18)

### Version of Immich Mobile App

N/A

### Platform with the issue

- [ ] Server
- [x] Web
- [ ] Mobile

### Device make and model

Desktop browser / local dev environment

### Your docker-compose.yml content

```YAML
Not applicable for this repro. The bug is in the web locale-selection logic and can be reproduced from a local checkout / unit test.
```

### Your .env content

```Shell
Not applicable for this repro.
```

### Reproduction steps

1.
2.
31. Use a web build that includes #27869.
32. 2. Set the stored language preference to an underscore-based locale that exists in `i18n/*.json`, for example `de_CH` or `zh_Hant`.
33. 3. Open the language settings / language selector.
34. 4. Observe that the selected language resolves to the English fallback instead of the stored locale.
Equivalent logic-level repro from current code:
1. `langs` stores values like `de_CH`.
2. 2. `langCodes` stores values like `de-CH`.
3. 3. `getClosestAvailableLocale(['de_CH'], langCodes)` returns `undefined`.
...


### Relevant log output

```shell
Focused local web regression test on current main:

FAIL  src/lib/i18n-locale-regression.autovalidation.spec.ts
AssertionError: expected undefined to be 'de_CH'
AssertionError: expected undefined to be 'pt_BR'
AssertionError: expected undefined to be 'sr_Cyrl'
AssertionError: expected undefined to be 'zh_Hant'
```

### Additional information

I validated this locally against the current `main` code path and reproduced the selector fallback effect with real locale files from the repo.

Representative result:
[
  {"stored_lang":"de_CH","closestLanguage":null,"selectedOptionValue":"en","falls_back_to_english":true},
  {"stored_lang":"pt_BR","closestLanguage":null,"selectedOptionValue":"en","falls_back_to_english":true},
  {"stored_lang":"sr_Cyrl","closestLanguage":null,"selectedOptionValue":"en","falls_back_to_english":true},
  {"stored_lang":"zh_Hant","closestLanguage":null,"selectedOptionValue":"en","falls_back_to_english":true}
]

This looks like a regression from #27869 / the broader #21337 work rather than an intended behavior change.

--- a/web/src/lib/i18n.spec.ts
+++ b/web/src/lib/i18n.spec.ts
@@ -50,5 +50,13 @@ describe('i18n', () => {
       expect(getClosestAvailableLocale(['zh'], allLocales)).toBeUndefined();
       expect(getClosestAvailableLocale(['de', 'zh', 'en-US'], allLocales)).toBe('en-US');
     });
+
+    it('matches underscore-based stored locale codes against normalized locale lists', () => {
+      const allLocales = ['de-CH', 'pt-BR', 'sr-Cyrl', 'zh-Hant'];
+      expect(getClosestAvailableLocale(['de_CH'], allLocales)).toBe('de_CH');
+      expect(getClosestAvailableLocale(['pt_BR'], allLocales)).toBe('pt_BR');
+      expect(getClosestAvailableLocale(['sr_Cyrl'], allLocales)).toBe('sr_Cyrl');
+      expect(getClosestAvailableLocale(['zh_Hant'], allLocales)).toBe('zh_Hant');
+    });
   });
 });

--- a/web/src/lib/utils/i18n.ts
+++ b/web/src/lib/utils/i18n.ts
@@ -19,21 +19,20 @@ const fileCodes = Object.keys(modules)
   .map((path) => path.match(/\/(\w+)\.json$/)?.[1])
   .filter(Boolean) as string[];
 
-const normalizeLocaleCode = (code: string) => code.replaceAll('_', '-');
-const convertBCP47 = (code: string) => normalizeLocaleCode(code);
+const convertBCP47 = (code: string) => code.replaceAll('_', '-');
 
 export const langCodes = fileCodes.map((code) => convertBCP47(code));
 
 // https://github.com/kaisermann/svelte-i18n/blob/780932a3e1270d521d348aac8ba03be9df309f04/src/runtime/stores/locale.ts#L11
 const getSubLocales = (locale: string) => {
-  return normalizeLocaleCode(locale)
+  return convertBCP47(locale)
     .split('-')
     .map((_, i, arr) => arr.slice(0, i + 1).join('-'))
     .reverse();
 };
 
 export const getClosestAvailableLocale = (locales: readonly string[], allLocales: readonly string[]) => {
-  const allLocalesSet = new Set(allLocales.map(normalizeLocaleCode));
+  const allLocalesSet = new Set(allLocales.map(convertBCP47));
   return locales.find((locale) => getSubLocales(locale).some((subLocale) => allLocalesSet.has(subLocale)));
 };
 

--- a/web/src/lib/utils/i18n.ts
+++ b/web/src/lib/utils/i18n.ts
@@ -32,7 +32,7 @@ const getSubLocales = (locale: string) => {
 };
 
 export const getClosestAvailableLocale = (locales: readonly string[], allLocales: readonly string[]) => {
-  const allLocalesSet = new Set(allLocales.map(convertBCP47));
+  const allLocalesSet = new Set(allLocales.map((locale) => convertBCP47(locale)));
   return locales.find((locale) => getSubLocales(locale).some((subLocale) => allLocalesSet.has(subLocale)));
 };
 

--- a/web/src/lib/utils/i18n.ts
+++ b/web/src/lib/utils/i18n.ts
@@ -19,20 +19,21 @@ const fileCodes = Object.keys(modules)
   .map((path) => path.match(/\/(\w+)\.json$/)?.[1])
   .filter(Boolean) as string[];
 
-const convertBCP47 = (code: string) => code.replace('_', '-');
+const normalizeLocaleCode = (code: string) => code.replaceAll('_', '-');
+const convertBCP47 = (code: string) => normalizeLocaleCode(code);
 
 export const langCodes = fileCodes.map((code) => convertBCP47(code));
 
 // https://github.com/kaisermann/svelte-i18n/blob/780932a3e1270d521d348aac8ba03be9df309f04/src/runtime/stores/locale.ts#L11
 const getSubLocales = (locale: string) => {
-  return locale
+  return normalizeLocaleCode(locale)
     .split('-')
     .map((_, i, arr) => arr.slice(0, i + 1).join('-'))
     .reverse();
 };
 
 export const getClosestAvailableLocale = (locales: readonly string[], allLocales: readonly string[]) => {
-  const allLocalesSet = new Set(allLocales);
+  const allLocalesSet = new Set(allLocales.map(normalizeLocaleCode));
   return locales.find((locale) => getSubLocales(locale).some((subLocale) => allLocalesSet.has(subLocale)));
 };
 

